packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,375 @@
+
+  /** Temporary variables. */
+  private static _tempRay: Ray = new Ray();
+  private static _tempPoint: Vector2 = new Vector2();
+  private static _tempHitResult: HitResult = new HitResult();
+
+  /** Whether to support multi-touch. */
+  private _multiTouchEnabled: boolean = false;
+  /** Number of scripts registered to listen. */
+  private _regScriptCount: number = 0;
+
+  /**
+   * Whether to support multi-touch.
+   */
+  get multiTouchEnabled(): boolean {
+    return this._multiTouchEnabled;
+  }
+
+  set multiTouchEnabled(enabled: boolean) {
+    this._multiTouchEnabled = enabled;
+  }
+
+  /**
+   * Constructor an InputManager.
+   * @param engine - The current engine instance
+   */
+  constructor(engine: Engine) {
+    // @ts-ignore
+    this._canvas = engine.canvas._webCanvas;
+    this._sceneMgr = engine.sceneManager;
+    this._physicsMgr = engine.physicsManager;
+  }
+
+  /**
+   *	Get touch pointer.
+   * 	@param	idx - Index of touch pointer
+   * 	@return Touch pointer
+   *  @remarks The returned Pointer should be considered deep-read-only.
+   */
+  getTouch(idx: number): Readonly<Pointer> {
+    return idx < this._actPointerCount ? this._pointerList[idx] : null;
+  }
+
+  /**
+   *	Get input.
+   * 	@return Input
+   *  @remarks The returned Input should be considered deep-read-only.
+   */
+  getInput(): Readonly<Input> {
+    return this._input;
+  }
+
+  /**
+   * Add the script that listen to input.
+   */
+  on() {
+    if (this._regScriptCount++ == 0) {
+      this._updateListener();
+    }
+  }
+
+  /**
+   * Remove the script that listen to input.
+   */
+  off() {
+    if (--this._regScriptCount == 0) {
+      this._updateListener();
+    }
+  }
+
+  /**
+   * Called when the engine is destroyed.
+   */
+  destroy() {
+    const { _canvas: canvas } = this;
+    canvas.removeEventListener("pointerdown", this._onPointerDown);
+    canvas.removeEventListener("pointerup", this._onPointerUp);
+    canvas.removeEventListener("pointermove", this._onPointerMove);
+    canvas.removeEventListener("pointercancel", this._onPointerCancelOrOut);
+    canvas.removeEventListener("pointerout", this._onPointerCancelOrOut);
+    this._eventList.length = 0;
+    this._pointerList.length = 0;
+    this._input = null;
+    this._pointerIdToIndex = null;
+    this._sceneMgr = null;
+    this._physicsMgr = null;
+    this._canvas = null;
+  }
+
+  /**
+   * Update pointer event, will be executed every frame.
+   * @internal
+   */
+  _update() {
+    const { _eventLen } = this;
+    if (_eventLen > 0 || this._actPointerCount > 0) {
+      const { _eventList, _input } = this;
+      let prePressedEntity = _input.pressedEntity;
+      if (prePressedEntity) {
+        /** Pointer Drag. */
+        const preScripts = prePressedEntity._scripts._elements;
+        for (let i = preScripts.length - 1; i >= 0; i--) {
+          preScripts[i].onPointerDrag();
+        }
+      }
+      /** Get the entity hit by the ray. */
+      /** @ts-ignore */
```

Comment:
If `canvas` is `HTMLCanvasElement`, it should not use `@ts-ignore`.

Reply:
Done.